File: app/src/main/java/com/example/islamic/adapter/Surah_Adapter.java
package com.example.islamic.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.islamic.R;
import com.example.islamic.listener.SurahListener;

import java.util.ArrayList;

public class Surah_Adapter extends RecyclerView.Adapter<Surah_Adapter.viewHolder> {
    ArrayList<Integer>index=new ArrayList<>();
    ArrayList<String>name=new ArrayList<>();
    private SurahListener surahListener;
    Context context ;

    public Surah_Adapter(ArrayList<Integer> index, ArrayList<String> name, SurahListener surahListener, Context context) {
        this.index = index;
        this.name = name;
        this.surahListener = surahListener;
        this.context = context;
    }

    @NonNull
    @Override
    public viewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view= LayoutInflater.from(parent.getContext()).inflate(R.layout.surah,null,false);
        return new viewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull viewHolder holder, int position) {
        if (holder.textView1 != null && holder.textView2 != null && name.get(position) != null && index.get(position) != null) {
            holder.textView1.setText(index.get(position).toString());
            holder.textView2.setText(name.get(position));
        }
    }

    @Override
    public int getItemCount() {
        return index.size();
    }

    public class viewHolder extends RecyclerView.ViewHolder {
        TextView textView1,textView2;
        public viewHolder(@NonNull View itemView) {
            super(itemView);
            textView1=itemView.findViewById(R.id.surah_number);
            textView2=itemView.findViewById(R.id.arabic_name);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
             surahListener.onSurahListener(getAdapterPosition());
                }
            });
        }
    }
}

File: app/src/main/java/com/eslamdev/islamic/presentation/ui/adapter/SurahAdapter.kt
package com.eslamdev.islamic.presentation.ui.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.eslamdev.islamic.data.model.SurahModel
import com.eslamdev.islamic.databinding.SurahItemForTafseerBinding
// ^ ملاحظة: هنستخدم نفس تصميم الـ XML القديم لأنه أشيك (surah_item_for_tafseer.xml)
// تأكد إن اسم الملف XML هو surah_item_for_tafseer.xml عشان الـ Binding ده يشتغل

class SurahAdapter(private val onClick: (SurahModel) -> Unit) :
    ListAdapter<SurahModel, SurahAdapter.SurahViewHolder>(DiffCallback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SurahViewHolder {
        val binding = SurahItemForTafseerBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return SurahViewHolder(binding)
    }

    override fun onBindViewHolder(holder: SurahViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item)
    }

    inner class SurahViewHolder(private val binding: SurahItemForTafseerBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(surah: SurahModel) {
            binding.surahNumber.text = surah.id.toString()
            binding.arabicName.text = surah.name

            binding.root.setOnClickListener {
                onClick(surah)
            }
        }
    }

    companion object DiffCallback : DiffUtil.ItemCallback<SurahModel>() {
        override fun areItemsTheSame(oldItem: SurahModel, newItem: SurahModel) = oldItem.id == newItem.id
        override fun areContentsTheSame(oldItem: SurahModel, newItem: SurahModel) = oldItem == newItem
    }
}